pkg/ctrl/doc.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package ctrl provides libraries for implementing Controllers.
+
+Controller
```

Comment:
Random thought: shouldn't this be the documentation for the "Controller" struct, rather than in the top level doc part? Maybe true also for some of the sections you've created below. Though I also understand why you're doing it this way.

Reply:
Hm.  I can put it there too.  There is a chicken/egg problem where users need to know what a Controller is before they will search it out.

Reply:
Maybe keep a brief introduction of how things are supposed to work together, so that people can then go below to get more information. Like...

> This package allows you to create and register your controllers so that you can reconcile blabla ...
> See example below.
>
> This example shows you how to register and start the reconcile loop:
> Example_Start() { ... }
>
> This example shows you how to set-up a more complicated Controller watcher ...
> Example_CombinesManyConcepts()

maybe?

Reply:
PTAL and let me know what you think.  I've changed the naming and docs quite a bit to address what are common miss conceptions about how controllers work.  The mechanics should be much more clear now.